iped-engine/src/main/java/iped/engine/data/MultiBitmapBookmarks.java

```diff
@@ -0,0 +1,529 @@
+                if (cnt != null) {
+                    ret += cnt;
+                }
+            }
+        }
+        return ret;
+    }
+
+    public void setInReport(String bookmarkName, boolean inReport) {
+        for (IBookmarks m : map.values()) {
+            int bookmarkId = m.getBookmarkId(bookmarkName);
+            m.setInReport(bookmarkId, inReport);
+        }
+    }
+
+    public boolean isInReport(String bookmarkName) {
+        for (IBookmarks m : map.values()) {
+            int bookmarkId = m.getBookmarkId(bookmarkName);
+            if (m.isInReport(bookmarkId))
+                return true;
+        }
+        return false;
+    }
+
+    public TreeSet<String> getBookmarkSet() {
+        TreeSet<String> bookmarks = new TreeSet<String>();
+        for (IBookmarks m : map.values())
+            bookmarks.addAll(m.getBookmarkMap().values());
+        return bookmarks;
+    }
+
+    /**
+     * return union of RoaringBitmap arrays with itemIds for each case of the
+     * bookmarks in parameter
+     */
+    public RoaringBitmap[] getBookmarksUnions(Set<String> bookmarkNames) {
+        SoftReference<RoaringBitmap[]> sbitmap = bsetCache.get(bookmarkNames.hashCode());
+        RoaringBitmap[] unions = null;
+        if (sbitmap != null) {
+            unions = sbitmap.get();
+        }
+        if (unions == null) {
+            unions = new RoaringBitmap[map.size()];
+
+            for (Entry<Integer, IBookmarks> e : map.entrySet()) {
+                IBookmarks bookmarks = (IBookmarks) e.getValue();
+                if (bookmarks instanceof BitmapBookmarks) {
+                    RoaringBitmap lunion = ((BitmapBookmarks) e.getValue()).getBookmarksUnion(bookmarkNames);
+                    unions[e.getKey()] = lunion;
+                }
+            }
+            bsetCache.put(bookmarkNames.hashCode(), new SoftReference<RoaringBitmap[]>(unions));
+        }
+
+        unions = unions.clone();
+        for (int i = 0; i < unions.length; i++) {
+            unions[i] = unions[i].clone();
+        }
+
+        return unions;
+    }
+
+    /**
+     * return union of RoaringBitmap arrays with itemIds for each case
+     */
+    public RoaringBitmap[] getBookmarksUnions() {
+        RoaringBitmap[] unions = unionsCache != null ? unionsCache.get() : null;
+        if (unions == null) {
+            unions = new RoaringBitmap[map.size()];
+
+            for (Entry<Integer, IBookmarks> e : map.entrySet()) {
+                IBookmarks bookmarks = (IBookmarks) e.getValue();
+                if (bookmarks instanceof BitmapBookmarks) {
+                    RoaringBitmap lunion = ((BitmapBookmarks) e.getValue()).getBookmarksUnion();
+                    unions[e.getKey()] = lunion;
+                }
+            }
+            unionsCache = new SoftReference<RoaringBitmap[]>(unions);
+        }
+        unions = unions.clone();
+        for (int i = 0; i < unions.length; i++) {
+            unions[i] = unions[i].clone();
+        }
```

Comment:
@patrickdalla are these SoftReferences needed? I thought RoaringBitmap operations should be fast...

Anyway, I think using a Set<String> hashcode as a SoftReference key is not a good idea, since it may lead to collisions. And are those clone calls needed? That clone loop looks unecessary to me, not sure.

Reply:
Sorry, I just saw RoaringBitmap operations are done in place, so cloning seems needed. But I changed the key used in the bsetCache SoftReferences Map from hashCode to a String built from the concatenation of all bookmark names.